contracts/TokenHandler.sol

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { ITokenHandler } from './interfaces/ITokenHandler.sol';
+import { IERC20 } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IERC20.sol';
+import { SafeTokenTransferFrom, SafeTokenCall } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/libs/SafeTransfer.sol';
+import { ReentrancyGuard } from '@axelar-network/axelar-gmp-sdk-solidity/contracts/utils/ReentrancyGuard.sol';
+
+import { IERC20MintableBurnable } from './interfaces/IERC20MintableBurnable.sol';
+import { IERC20BurnableFrom } from './interfaces/IERC20BurnableFrom.sol';
+
+/**
+ * @title ITokenManager Interface
+ * @notice This interface is responsible for handling tokens before initiating an interchain token transfer, or after receiving one.
+ */
+contract TokenHandler is ITokenHandler, ReentrancyGuard {
+    using SafeTokenTransferFrom for IERC20;
+    using SafeTokenCall for IERC20;
+
+    /**
+     * @notice This function gives token to a specified address.
+     * @dev Can only be called by the service.
+     * @param tokenManagerType The token manager type.
+     * @param tokenAddress the address of the token to give.
+     * @param tokenManager the address of the token manager.
+     * @param to the address of the recepient.
+     * @return amount The amount of tokens actually given, which will only be different than `amount` in cases where the token takes some on-transfer fee.
+     */
+    // slither-disable-next-line locked-ether
+    function giveToken(
+        uint256 tokenManagerType,
+        address tokenAddress,
+        address tokenManager,
+        address to,
+        uint256 amount
+    ) external payable noReEntrancy returns (uint256) {
```

Comment:
Can we apply the re entrancy guard only for certain token types instead? Might be easier if we have separate internal methods and we can apply it on those?

Reply:
only applying to lockUnlockFee now